### PR TITLE
code to create chart without reading template file in PPT, Test Example file for DOC and PPT

### PR DESCRIPTION
**Pull request by sandeeptiwari32**
added code to create chart without reading template file in PPT and Test Example file for DOC and PPT

### CODE REVIEW

File: src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java [hunk 1]
```diff
@@ -81,6 +82,27 @@
 
 @Beta
 public abstract class XDDFChart extends POIXMLDocumentPart implements TextContainer {
+
```
Comment (sandeeptiwari32) [inline src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java:85]: moved constant variable code from XWPFChart TO XDDFChart

**Comment by asfgit** [comment]
Can one of the admins verify this patch?

File: src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java [hunk 2]
```diff
@@ -712,10 +734,29 @@ public CellReference setSheetTitle(String title, int column) {
         XSSFRow row = this.getRow(sheet, 0);
         XSSFCell cell = this.getCell(row, column);
         cell.setCellValue(title);
-        this.updateSheetTable(sheet.getTables().get(0).getCTTable(), title, column);
+
```
Comment (sandeeptiwari32) [inline src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java:737]: if sheet does't have table object then first create it and then pass newly created table to update sheet table data. so for that purpose created a method GetSheetTable to return table object.

File: src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java [hunk 3]
```diff
@@ -712,10 +734,29 @@ public CellReference setSheetTitle(String title, int column) {
...
         cell.setCellValue(title);
-        this.updateSheetTable(sheet.getTables().get(0).getCTTable(), title, column);
+        
+        CTTable ctTable = this.getSheetTable(sheet);
+        
+        this.updateSheetTable(ctTable, title, column);
         return new CellReference(sheet.getSheetName(), 0, column, true, true);
     }
 
+    /**
```
Comment (sandeeptiwari32) [inline src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java:744]: method which check whether sheet have table object, in case table size is 0 then create table and return that table.

File: src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java [hunk 4]
```diff
@@ -729,9 +770,11 @@ public CellReference setSheetTitle(String title, int column) {
     private void updateSheetTable(CTTable ctTable, String title, int index) {
         CTTableColumns tableColumnList = ctTable.getTableColumns();
         CTTableColumn column = null;
```
Comment (sandeeptiwari32) [inline src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java]: instead of starting loop  form 0 each time start it with table column size. previously each time we are adding new column from 0 inseatd from required index.

File: src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java [hunk 5]
```diff
@@ -729,9 +770,11 @@ public CellReference setSheetTitle(String title, int column) {
...
-        for( int i = 0; tableColumnList.getCount() < index; i++) {
+        int columnCount  = tableColumnList.getTableColumnList().size()-1;
```
Comment (sandeeptiwari32) [inline src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java:773]: instead of starting loop  form 0 each time start it with table column size. previously each time we are adding new column from 0 inseatd from required index.

File: src/ooxml/java/org/apache/poi/xslf/usermodel/XMLSlideShow.java [hunk 6]
```diff
@@ -302,14 +302,23 @@ public XSLFSlide createSlide() {
      * Create a blank chart on the given slide.
      */
     public XSLFChart createChart(XSLFSlide slide) {
+        XSLFChart chart = createChart();
+        slide.addRelation(null, XSLFRelation.CHART, chart);
+        return chart;
+    }
+    
+    /**
```
Comment (sandeeptiwari32) [inline src/ooxml/java/org/apache/poi/xslf/usermodel/XMLSlideShow.java:310]: added a method to create template for chart as we have method for XWPFDocument and XSSFWorkbook.

File: src/ooxml/java/org/apache/poi/xslf/usermodel/XMLSlideShow.java [hunk 7]
```diff
@@ -416,7 +425,7 @@ public XSLFNotesMaster getNotesMaster() {
      * Return all the charts in the slideshow
      */
     public List<XSLFChart> getCharts() {
-        return _charts;
+        return Collections.unmodifiableList(_charts);
```
Comment (sandeeptiwari32) [inline src/ooxml/java/org/apache/poi/xslf/usermodel/XMLSlideShow.java:428]: changed it to unmodifiable list as we have unmodifiable list in XWPFDocument and XSSFWorkbook.

File: src/ooxml/java/org/apache/poi/xslf/usermodel/XSLFChart.java [hunk 8]
```diff
@@ -90,4 +104,46 @@ protected CTTextBody getTextBody(boolean create) {
             };
         }
     }
+
```
Comment (sandeeptiwari32) [inline src/ooxml/java/org/apache/poi/xslf/usermodel/XSLFChart.java:107]: method to create graphic frame for chart object in XSLFSLIDE

File: src/ooxml/java/org/apache/poi/xslf/usermodel/XSLFDrawing.java [hunk 9]
```diff
@@ -106,6 +106,18 @@ public XSLFTable createTable(){
         shape.setAnchor(new Rectangle2D.Double());
         return shape;
     }
+
```
Comment (sandeeptiwari32) [inline src/ooxml/java/org/apache/poi/xslf/usermodel/XSLFDrawing.java:109]: method to add chart into graphic frame of slide

File: src/ooxml/java/org/apache/poi/xslf/usermodel/XSLFSheet.java [hunk 10]
```diff
@@ -718,5 +720,32 @@ public XSLFPlaceholderDetails getPlaceholderDetails(Placeholder placeholder) {
         final XSLFSimpleShape ph = getPlaceholder(placeholder);
         return (ph == null) ? null : new XSLFPlaceholderDetails(ph);
     }
+    
+    /**
+     * this method will add chart into slide
```
Comment (sandeeptiwari32) [inline src/ooxml/java/org/apache/poi/xslf/usermodel/XSLFSheet.java:725]: method to add chart into slide with default height width and co-ordinate.

File: src/ooxml/java/org/apache/poi/xslf/usermodel/XSLFSheet.java [hunk 11]
```diff
@@ -718,5 +720,32 @@ public XSLFPlaceholderDetails getPlaceholderDetails(Placeholder placeholder) {
...
+     */
+    public XSLFChart addChart(XSLFChart chart)
+    {
+        Rectangle2D rect2D = new java.awt.Rectangle(XDDFChart.DEFAULT_X, XDDFChart.DEFAULT_Y, 
+                XDDFChart.DEFAULT_WIDTH, XDDFChart.DEFAULT_HEIGHT);
+        
+        return addChart(chart, rect2D);
+    }
+    
+    /**
```
Comment (sandeeptiwari32) [inline src/ooxml/java/org/apache/poi/xslf/usermodel/XSLFSheet.java:738]: method to add chart into slide with user defined height width and co-ordinate.

File: src/ooxml/java/org/apache/poi/xwpf/usermodel/XWPFChart.java [hunk 12]
```diff
@@ -36,16 +36,7 @@ Licensed to the Apache Software Foundation (ASF) under one or more
...
-     * default width of chart in emu
-     */
-    public static final int DEFAULT_WIDTH = 500000;
-
-    /**
-     * default height of chart in emu
-     */
-    public static final int DEFAULT_HEIGHT = 500000;
-
+
```
Comment (sandeeptiwari32) [inline src/ooxml/java/org/apache/poi/xwpf/usermodel/XWPFChart.java]: moved code into XDDFChart class

File: src/ooxml/java/org/apache/poi/xwpf/usermodel/XWPFDocument.java [hunk 13]
```diff
@@ -1672,7 +1673,7 @@ public XWPFDocument getXWPFDocument() {
      * @since POI 4.0.0
      */
     public XWPFChart createChart() throws InvalidFormatException, IOException {
-        return createChart(XWPFChart.DEFAULT_WIDTH, XWPFChart.DEFAULT_HEIGHT);
+        return createChart(XDDFChart.DEFAULT_WIDTH, XDDFChart.DEFAULT_HEIGHT);
```
Comment (sandeeptiwari32) [inline src/ooxml/java/org/apache/poi/xwpf/usermodel/XWPFDocument.java:1676]: changed constant class name

File: src/ooxml/testcases/org/apache/poi/xslf/usermodel/TestXSLFSlide.java [hunk 14]
```diff
@@ -194,4 +195,16 @@ public void testMergeSlides() throws IOException {
         
         ppt.close();
     }    
+    
+    @Test
```
Comment (sandeeptiwari32) [inline src/ooxml/testcases/org/apache/poi/xslf/usermodel/TestXSLFSlide.java:199]: junit test case for new method in XMLSlideShow "createChart"

**Review by sandeeptiwari32 — COMMENTED** [review]
code to create chart without reading template file for PPT

**Comment by sandeeptiwari32** [comment]
code to create chart without reading template file for PPT

File: src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java [hunk 15]
```diff
@@ -729,9 +770,11 @@ public CellReference setSheetTitle(String title, int column) {
...
+        for( int i = columnCount; columnCount < index; i++) {
             column = tableColumnList.addNewTableColumn();
             column.setId(i);
+            ++columnCount;
         }
```
Comment (pjfanning) [inline src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java:777]: Could you use this?
```
for (int i = columnCount; i < index; i++) {
    column = tableColumnList.addNewTableColumn();
    column.setId(i);
}
```

Comment (sandeeptiwari32) [reply src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java:777]: my bad.. :P 
done required change.

File: src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java (hunk 1 above)
Comment (pjfanning) [reply src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java:85]: with the moved constants, we can't delete them from XWPFChart as it would cause issues for other users but could you set the XWPFChart values to be be based on the XDDF ones?

```
public static final int DEFAULT_WIDTH = XDDFChart.DEFAULT_WIDTH; 
```

File: src/examples/src/org/apache/poi/xwpf/usermodel/examples/AutoChartDOCExample.java [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+        		listSpeakers.add(Double.valueOf(vals[1]));
+        		listLanguages.add(vals[2]);
+        	}
+        	
+        	String[] categories = listLanguages.toArray(new String[listLanguages.size()]);
+        	Double[] values1 = listCountries.toArray(new Double[listCountries.size()]);
+        	Double[] values2 = listSpeakers.toArray(new Double[listSpeakers.size()]);
+
+        	try (XWPFDocument doc = new XWPFDocument()) {
+        		XWPFChart chart = doc.createChart(5000000,5000000);
```
Comment (Alain-Bearez) [inline src/examples/src/org/apache/poi/xwpf/usermodel/examples/AutoChartDOCExample.java]: Please use `XDDFChart.DEFAULT_WIDTH` and `XDDFChart.DEFAULT_HEIGHT` instead of these magic numbers.

File: src/examples/src/org/apache/poi/xslf/usermodel/AutoChartPPTExample.java [hunk 17]
```diff
@@ -0,0 +1,148 @@
...
+        	Double[] values1 = listCountries.toArray(new Double[listCountries.size()]);
+        	Double[] values2 = listSpeakers.toArray(new Double[listSpeakers.size()]);
+
+        	try {
+        		
+        		XMLSlideShow ppt = new XMLSlideShow();
+        		XSLFSlide slide = ppt.createSlide();
+        		XSLFChart chart = ppt.createChart();
+        		Rectangle2D rect2D = new java.awt.Rectangle(XDDFChart.DEFAULT_X, XDDFChart.DEFAULT_Y, 
+        				5000000, 5000000);
```
Comment (Alain-Bearez) [inline src/examples/src/org/apache/poi/xslf/usermodel/AutoChartPPTExample.java]: Please use `XDDFChart.DEFAULT_WIDTH` and `XDDFChart.DEFAULT_HEIGHT` instead of these magic numbers.

File: src/ooxml/java/org/apache/poi/xslf/usermodel/XMLSlideShow.java [hunk 18]
```diff
@@ -302,14 +302,23 @@ public XSLFSlide createSlide() {
...
     public XSLFChart createChart(XSLFSlide slide) {
+        XSLFChart chart = createChart();
+        slide.addRelation(null, XSLFRelation.CHART, chart);
+        return chart;
+    }
+    
+    /**
+     * This method is used to create template for chart XML.
+     * @return Xslf chart object 
+     */
```
Comment (Alain-Bearez) [inline src/ooxml/java/org/apache/poi/xslf/usermodel/XMLSlideShow.java:314]: add the `@since 4.0.2` as we might not be able to ship it with currently reviewed `4.0.1` to be released after the week-end

File: src/ooxml/java/org/apache/poi/xslf/usermodel/XMLSlideShow.java [hunk 19]
```diff
@@ -302,14 +302,23 @@ public XSLFSlide createSlide() {
...
+    public XSLFChart createChart() {
         int chartIdx = findNextAvailableFileNameIndex(XSLFRelation.CHART, _charts.size() + 1);
         XSLFChart chart = (XSLFChart) createRelationship(XSLFRelation.CHART, XSLFFactory.getInstance(), chartIdx, true).getDocumentPart();
-        slide.addRelation(null, XSLFRelation.CHART, chart);
         chart.setChartIndex(chartIdx);
         _charts.add(chart);
         return chart;
     }
-
+
```
Comment (Alain-Bearez) [inline src/ooxml/java/org/apache/poi/xslf/usermodel/XMLSlideShow.java]: no trailing whitespaces, please

File: src/ooxml/java/org/apache/poi/xslf/usermodel/XMLSlideShow.java (hunk 7 above)
Comment (Alain-Bearez) [reply src/ooxml/java/org/apache/poi/xslf/usermodel/XMLSlideShow.java:428]: great catch!

File: src/ooxml/java/org/apache/poi/xslf/usermodel/XSLFChart.java [hunk 20]
```diff
@@ -19,23 +19,37 @@
...
+import org.openxmlformats.schemas.presentationml.x2006.main.CTGraphicalObjectFrame;
+import org.openxmlformats.schemas.presentationml.x2006.main.CTGraphicalObjectFrameNonVisual;
 
 /**
  * Represents a Chart in a .pptx presentation
  */
 @Beta
 public final class XSLFChart extends XDDFChart {
 
+    static String CHART_URI = "http://schemas.openxmlformats.org/drawingml/2006/chart";
```
Comment (Alain-Bearez) [inline src/ooxml/java/org/apache/poi/xslf/usermodel/XSLFChart.java]: Does this really need to be of `package` visibility? Wouldn't `private` be OK?

File: src/ooxml/java/org/apache/poi/xslf/usermodel/XSLFChart.java [hunk 21]
```diff
@@ -90,4 +104,46 @@ protected CTTextBody getTextBody(boolean create) {
...
+    
+    /**
+     * method to add graphic frame for XSLF chart
+     *
+     * @param shapeId shape id
+     * @param rID relation id
+     * @param anchor size and location of chart
+     * @return graphic frame object
+     */
+    static CTGraphicalObjectFrame prototype(int shapeId, String rID, Rectangle2D anchor){
```
Comment (Alain-Bearez) [inline src/ooxml/java/org/apache/poi/xslf/usermodel/XSLFChart.java:117]: Does this really need to be of `package` visibility? Wouldn't `private` be OK?

File: src/ooxml/java/org/apache/poi/xslf/usermodel/XSLFDrawing.java [hunk 22]
```diff
@@ -106,6 +106,18 @@ public XSLFTable createTable(){
         shape.setAnchor(new Rectangle2D.Double());
         return shape;
     }
+    
+    /**
+     * This method will add chart into slide's graphic frame
+     *
+     * @param rID relation id of chart
+     * @param rect2D Chart Bounding values
+     */
```
Comment (Alain-Bearez) [inline src/ooxml/java/org/apache/poi/xslf/usermodel/XSLFDrawing.java:116]: add the `@since 4.0.2` as we might not be able to ship it with currently reviewed `4.0.1` to be released after the week-end

File: src/ooxml/java/org/apache/poi/xslf/usermodel/XSLFSheet.java [hunk 23]
```diff
@@ -718,5 +720,32 @@ public XSLFPlaceholderDetails getPlaceholderDetails(Placeholder placeholder) {
         final XSLFSimpleShape ph = getPlaceholder(placeholder);
         return (ph == null) ? null : new XSLFPlaceholderDetails(ph);
     }
+    
+    /**
+     * this method will add chart into slide
+     * with default height, width, x and y
+     * @param chart xslf chart object
+     * @return xslf chart object
+     */
```
Comment (Alain-Bearez) [inline src/ooxml/java/org/apache/poi/xslf/usermodel/XSLFSheet.java:729]: add the `@since 4.0.2` as we might not be able to ship it with currently reviewed `4.0.1` to be released after the week-end

File: src/ooxml/java/org/apache/poi/xwpf/usermodel/XWPFChart.java (hunk 12 above)
Comment (Alain-Bearez) [reply src/ooxml/java/org/apache/poi/xwpf/usermodel/XWPFChart.java]: Like already mentioned in other review, please refrain from deleting `public` code that other people might already be using in their production code.

Keep the definition of these constants as being `= XDDFChart.DEFAULT_WIDTH` and  `= XDDFChart.DEFAULT_HEIGHT`

File: src/ooxml/testcases/org/apache/poi/xslf/usermodel/TestXSLFSlide.java [hunk 24]
```diff
@@ -194,4 +195,16 @@ public void testMergeSlides() throws IOException {
...
+    public void testAutoChart() throws IOException {
+        XMLSlideShow ppt = new XMLSlideShow();
+        XSLFSlide slide = ppt.createSlide();
+        XSLFChart chart = ppt.createChart();
+        slide.addChart(chart);
+        
+        assertNotNull(chart);
```
Comment (Alain-Bearez) [inline src/ooxml/testcases/org/apache/poi/xslf/usermodel/TestXSLFSlide.java]: I would have checked that the returned chart from `slide.addChart` is the one you expect to be, i.e. the original `chart` passed as argument.

**Review by Alain-Bearez — CHANGES_REQUESTED** [review]
I am so glad to see such a useful contribution complementing my original use case!

File: src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java (hunk 1 above)
Comment (sandeeptiwari32) [reply src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java:85]: got it...
merged required code.

File: src/ooxml/java/org/apache/poi/xslf/usermodel/XMLSlideShow.java (hunk 7 above)
Comment (sandeeptiwari32) [reply src/ooxml/java/org/apache/poi/xslf/usermodel/XMLSlideShow.java:428]: thank you Alain Sir... :)

File: src/ooxml/java/org/apache/poi/xwpf/usermodel/XWPFChart.java (hunk 12 above)
Comment (sandeeptiwari32) [reply src/ooxml/java/org/apache/poi/xwpf/usermodel/XWPFChart.java]: merged required code.

File: src/examples/src/org/apache/poi/xwpf/usermodel/examples/AutoChartDOCExample.java (hunk 16 above)
Comment (sandeeptiwari32) [reply src/examples/src/org/apache/poi/xwpf/usermodel/examples/AutoChartDOCExample.java]: chnaged hard coded value to constant.

File: src/examples/src/org/apache/poi/xslf/usermodel/AutoChartPPTExample.java (hunk 17 above)
Comment (sandeeptiwari32) [reply src/examples/src/org/apache/poi/xslf/usermodel/AutoChartPPTExample.java]: chnaged hard coded value to constant

File: src/ooxml/java/org/apache/poi/xslf/usermodel/XMLSlideShow.java (hunk 18 above)
Comment (sandeeptiwari32) [reply src/ooxml/java/org/apache/poi/xslf/usermodel/XMLSlideShow.java:314]: Added Tag POI 4.0.2

File: src/ooxml/java/org/apache/poi/xslf/usermodel/XMLSlideShow.java (hunk 19 above)
Comment (sandeeptiwari32) [reply src/ooxml/java/org/apache/poi/xslf/usermodel/XMLSlideShow.java]: removed space.

File: src/ooxml/java/org/apache/poi/xslf/usermodel/XSLFChart.java (hunk 20 above)
Comment (sandeeptiwari32) [reply src/ooxml/java/org/apache/poi/xslf/usermodel/XSLFChart.java]: chnaged to private.

File: src/ooxml/java/org/apache/poi/xslf/usermodel/XSLFChart.java (hunk 21 above)
Comment (sandeeptiwari32) [reply src/ooxml/java/org/apache/poi/xslf/usermodel/XSLFChart.java:117]: we are using this method from XSLF drawing as we are using for other shap controls. so need protected or package visibility.

File: src/ooxml/java/org/apache/poi/xslf/usermodel/XSLFDrawing.java (hunk 22 above)
Comment (sandeeptiwari32) [reply src/ooxml/java/org/apache/poi/xslf/usermodel/XSLFDrawing.java:116]: Added Tag POI 4.0.2

File: src/ooxml/java/org/apache/poi/xslf/usermodel/XSLFSheet.java (hunk 23 above)
Comment (sandeeptiwari32) [reply src/ooxml/java/org/apache/poi/xslf/usermodel/XSLFSheet.java:729]: Added Tag POI 4.0.2

File: src/ooxml/testcases/org/apache/poi/xslf/usermodel/TestXSLFSlide.java (hunk 24 above)
Comment (sandeeptiwari32) [reply src/ooxml/testcases/org/apache/poi/xslf/usermodel/TestXSLFSlide.java]: changed method slide.addChart return type. no need to return XSLFchart object. we are using XSLFChart object to add relationship into slide.

**Comment by sandeeptiwari32** [comment]
added required changes.

File: src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java (hunk 1 above)
Comment (sandeeptiwari32) [reply src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java:85]: merged required code.

File: src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java [hunk 25]
```diff
@@ -712,10 +734,30 @@ public CellReference setSheetTitle(String title, int column) {
...
+        this.updateSheetTable(ctTable, title, column);
         return new CellReference(sheet.getSheetName(), 0, column, true, true);
     }
 
+    /**
+     * this method will check whether sheet have table
+     * in case table size zero then create new table and add table columns element
+     * @param sheet
+     * @return table object
+     * @since POI 4.0.2
```
Comment (Alain-Bearez) [inline src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java]: There is no need for `@since` tags on private methods.

File: src/ooxml/java/org/apache/poi/xwpf/usermodel/XWPFChart.java [hunk 26]
```diff
@@ -36,16 +36,16 @@ Licensed to the Apache Software Foundation (ASF) under one or more
  */
 @Beta
 public class XWPFChart extends XDDFChart {
-    /**
+   /**
```
Comment (Alain-Bearez) [inline src/ooxml/java/org/apache/poi/xwpf/usermodel/XWPFChart.java]: why the change in indentation here?

File: src/ooxml/java/org/apache/poi/xwpf/usermodel/XWPFChart.java [hunk 27]
```diff
@@ -36,16 +36,16 @@ Licensed to the Apache Software Foundation (ASF) under one or more
...
-    public static final int DEFAULT_WIDTH = 500000;
+    public static final int DEFAULT_WIDTH = XDDFChart.DEFAULT_WIDTH;
 
     /**
      * default height of chart in emu
      */
-    public static final int DEFAULT_HEIGHT = 500000;
-
+    public static final int DEFAULT_HEIGHT = XDDFChart.DEFAULT_HEIGHT;
+
```
Comment (Alain-Bearez) [inline src/ooxml/java/org/apache/poi/xwpf/usermodel/XWPFChart.java]: no trailing whitespace, please!

**Review by Alain-Bearez — COMMENTED** [review]
3 more details

File: src/ooxml/java/org/apache/poi/xwpf/usermodel/XWPFChart.java (hunk 26 above)
Comment (sandeeptiwari32) [reply src/ooxml/java/org/apache/poi/xwpf/usermodel/XWPFChart.java]: might be appeared because i edited it github web portal.
managed indentation.

File: src/ooxml/java/org/apache/poi/xwpf/usermodel/XWPFChart.java (hunk 27 above)
Comment (sandeeptiwari32) [reply src/ooxml/java/org/apache/poi/xwpf/usermodel/XWPFChart.java]: might be appeared because i edited it github web portal.
removed space.

File: src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java [hunk 28]
```diff
@@ -712,10 +734,29 @@ public CellReference setSheetTitle(String title, int column) {
...
+        
+        this.updateSheetTable(ctTable, title, column);
         return new CellReference(sheet.getSheetName(), 0, column, true, true);
     }
 
+    /**
+     * this method will check whether sheet have table
+     * in case table size zero then create new table and add table columns element
+     * @param sheet
+     * @return table object
```
Comment (sandeeptiwari32) [inline src/ooxml/java/org/apache/poi/xddf/usermodel/chart/XDDFChart.java:748]: removed since tag

File: src/ooxml/java/org/apache/poi/xwpf/usermodel/XWPFChart.java [hunk 29]
```diff
@@ -39,12 +39,12 @@ Licensed to the Apache Software Foundation (ASF) under one or more
     /**
```
Comment (sandeeptiwari32) [inline src/ooxml/java/org/apache/poi/xwpf/usermodel/XWPFChart.java:39]: might be appeared because i edited it github web portal.
managed indentation.

File: src/ooxml/java/org/apache/poi/xwpf/usermodel/XWPFChart.java [hunk 30]
```diff
@@ -39,12 +39,12 @@ Licensed to the Apache Software Foundation (ASF) under one or more
...
      * default width of chart in emu
      */
-    public static final int DEFAULT_WIDTH = 500000;
+    public static final int DEFAULT_WIDTH = XDDFChart.DEFAULT_WIDTH;
 
     /**
      * default height of chart in emu
      */
-    public static final int DEFAULT_HEIGHT = 500000;
+    public static final int DEFAULT_HEIGHT = XDDFChart.DEFAULT_HEIGHT;
```
Comment (sandeeptiwari32) [inline src/ooxml/java/org/apache/poi/xwpf/usermodel/XWPFChart.java:48]: might be appeared because i edited it github web portal.
removed space.

**Comment by sandeeptiwari32** [comment]
updated required changes.

**Comment by pjfanning** [comment]
@Alain-Bearez POI 4.0.1 is released. Do you think it is ok to merge this now?

**Comment by sandeeptiwari32** [comment]
@Alain-Bearez is it ok to merge current changes, or still need to change something.

**Review by Alain-Bearez — APPROVED** [review]

**Comment by Alain-Bearez** [comment]
I am still fixing some formatting minor issues before merging.

**Comment by Alain-Bearez** [comment]
merged by r1848432

**Comment by Alain-Bearez** [comment]
@pjfanning how could we make sure all the required new `CT*` classes on 
https://github.com/apache/poi/pull/135/files#diff-e7bf6663c76af8916ed221806ed98fa0
are present in the ooxml-lite?

Only then you could close this PR.

**Comment by pjfanning** [comment]
@sandeeptiwari32 can you close the PR? I ran a quick check and it looks like we have all the XMLBeans generated classes that we need in poi-ooxml-schemas jar. That jar is a subset of the full ooxml-schemas.jar and the classes in subset are derived from the classes that are used in our unit tests.

**Comment by sandeeptiwari32** [comment]
@pjfanning sure.

**Comment by sandeeptiwari32** [comment]
close pull request.

**Comment by 18876293672** [comment]
.......................................................................................
